File: python/questions/exist.py
from typing import List

class Solution:
    DIRS = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    def exist(self, board: List[List[str]], word: str) -> bool:
        rs = len(board)
        cs = len(board[0])
        ws = len(word)
        fg = [[False] * cs for _ in range(rs)]
        def fw(r: int, c: int, p: int) -> bool:
            if r<0 or r>=rs or c<0 or c>=cs or fg[r][c] or board[r][c]!=word[p]:
                return False
            if p == ws-1:
                return True
            fg[r][c] = True
            for dr, dc in self.DIRS:
                if fw(r+dr, c+dc, p+1):
                    return True
            fg[r][c] = False
            return False
        for i in range(rs):
            for j in range(cs):
                if fw(i, j, 0):
                    return True
        return False

File: rust/questions/exist.rs
impl Solution {

    const DIRS: [(isize, isize); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let rs = board.len();
        let cs = board[0].len();
        let wd: Vec<char> = word.chars().collect();
        let ws = wd.len();
        let mut fg = vec![vec![false; cs]; rs];
        fn fw(r: usize, c: usize, p: usize, rs: usize, cs: usize, ws: usize, bd: &Vec<Vec<char>>, wd: &[char], fg: &mut Vec<Vec<bool>>) -> bool {
            if bd[r][c] != wd[p] || fg[r][c] {
                return false;
            }
            if p == ws-1 {
                return true;
            }
            fg[r][c] = true;
            for (dr, dc) in Solution::DIRS {
                let nr = r as isize + dr;
                let nc = c as isize + dc;
                if nr>=0 && nc>=0 {
                    let vnr = nr as usize;
                    let vnc = nc as usize;
                    if vnr < rs && vnc < cs {
                        if fw(vnr, vnc, p+1, rs, cs, ws, bd, wd, fg) {
                            return true;
                        }
                    }
                }
            }
            fg[r][c] = false;
            return false;
        }
        for r in 0..rs {
            for c in 0..cs {
                if fw(r, c, 0, rs, cs, ws, &board, &wd, &mut fg) {
                    return true;
                }
            }
        }
        false
    }
}
